### PR TITLE
Update key related byron era cli commands with cardano-api

### PR DESCRIPTION
**Pull request by Jimbo4350**

### CODE REVIEW

File: cardano-api/src/Cardano/Api.hs [hunk 1]
```diff
@@ -422,6 +423,7 @@ module Cardano.Api (
     -- ** Conversions
     --TODO: arrange not to export these
     toNetworkMagic,
+    toByronLovelace,
```
Comment (dcoutts) [inline cardano-api/src/Cardano/Api.hs:426]: This should go via `Cardano.Api.Byron` not `Cardano.Api`

File: cardano-api/src/Cardano/Api/KeysByron.hs [hunk 2]
```diff
@@ -54,6 +68,127 @@ import           Cardano.Api.SerialiseTextEnvelope
...
+  ByronLegacyKeyFormat :: ByronKeyFormat ByronKeyLegacy
+  ByronNonLegacyKeyFormat :: ByronKeyFormat ByronKey
+
+instance IsByronLegacyFormat ByronKey where
+  isLegacyFormat = ByronNonLegacyKeyFormat
+
+instance IsByronLegacyFormat ByronKeyLegacy where
+  isLegacyFormat = ByronLegacyKeyFormat
+
+instance HasTypeProxy ByronKeyLegacy where
```
Comment (dcoutts) [inline cardano-api/src/Cardano/Api/KeysByron.hs:86]: We should probably do all the ByronKeyLegacy instances after ByronKey in this module, just for ease of reading. So yes the two tag data types (ByronKey and ByronKeyLegacy) first, then the classes and types that cover both (class IsByronKey and data ByronKeyFormat), then sections for the impl for the specific key types, ByronKey then ByronKeyLegacy. It's ok to use section separators like we do in other modules.

File: cardano-api/src/Cardano/Api/Byron.hs [hunk 3]
```diff
@@ -68,9 +68,12 @@ module Cardano.Api.Byron
     NetworkMagic(..),
 
     -- ** Conversions
+    fromByronTxIn,
     toByronNetworkMagic,
     toByronProtocolMagicId,
     toByronRequiresNetworkMagic,
+
+    makeByronTransaction,
```
Comment (dcoutts) [inline cardano-api/src/Cardano/Api/Byron.hs:76]: Wasn't this already exported via `Cardano.Api`?

File: cardano-api/src/Cardano/Api/KeysByron.hs [hunk 4]
```diff
@@ -54,6 +68,127 @@ import           Cardano.Api.SerialiseTextEnvelope
 -- This is a type level tag, used with other interfaces like 'Key'.
 --
 data ByronKey
+data ByronKeyLegacy
+
+class IsByronLegacyFormat key where
```
Comment (dcoutts) [inline cardano-api/src/Cardano/Api/KeysByron.hs:73]: Naming: later on you use this class to generalise from ByronKey to any Byron key. But the name `IsByronLegacyFormat key =>` suggests that key is a byron legacy format, but that's not what you mean: you mean it's a Byron key type, and the class can tell you if the format is legacy or modern.

So I'd suggest instead:
```
class IsByronKey key where
    byronKeyFormat :: ByronKeyFormat key
```

File: cardano-api/src/Cardano/Api/KeysByron.hs [hunk 5]
```diff
@@ -54,6 +68,127 @@ import           Cardano.Api.SerialiseTextEnvelope
...
+  isLegacyFormat :: ByronKeyFormat key
+
+data ByronKeyFormat key where
+  ByronLegacyKeyFormat :: ByronKeyFormat ByronKeyLegacy
+  ByronNonLegacyKeyFormat :: ByronKeyFormat ByronKey
```
Comment (dcoutts) [inline cardano-api/src/Cardano/Api/KeysByron.hs:78]: Naming: perhaps the opposite of "legacy" is "modern"? :-)

File: cardano-api/src/Cardano/Api/Tx.hs [hunk 6]
```diff
@@ -521,9 +521,10 @@ makeSignedTransaction witnesses (ShelleyTxBody era txbody txmetadata) =
           (maybeToStrictMaybe txmetadata)
 
 
-makeByronKeyWitness :: NetworkId
+makeByronKeyWitness :: forall key. IsByronLegacyFormat key
```
Comment (dcoutts) [inline cardano-api/src/Cardano/Api/Tx.hs:524]: See how this name is misleading? Hence the suggestion for `IsByronKey`

File: cardano-api/src/Cardano/Api/Tx.hs [hunk 7]
```diff
@@ -535,11 +536,18 @@ makeByronKeyWitness nw (ByronTxBody txbody) =
...
-     in \(ByronSigningKey sk) ->
-        ByronKeyWitness $
-          Byron.VKWitness
-            (Byron.toVerification sk)
-            (Byron.sign pm Byron.SignTx sk (Byron.TxSigData txhash))
+     in case isLegacyFormat :: ByronKeyFormat key of
+          ByronLegacyKeyFormat ->
+            \(ByronSigningKeyLegacy sk) -> witness sk pm txhash
+          ByronNonLegacyKeyFormat ->
+            \(ByronSigningKey sk) -> witness sk pm txhash
```
Comment (dcoutts) [inline cardano-api/src/Cardano/Api/Tx.hs:543]: :trophy:

File: cardano-cli/src/Cardano/CLI/Byron/Key.hs [hunk 8]
```diff
@@ -134,17 +122,11 @@ readPaymentVerificationKey (VerificationKeyFile fp) = do
...
-serialisePoorKey bKeyFormat ps =
-  serialiseSigningKey bKeyFormat $ Genesis.poorSecretToKey ps
-
 -- | Generate a cryptographically random signing key,
 --   protected with a (potentially empty) passphrase.
-keygen :: Crypto.PassPhrase -> IO SigningKey
+keygen :: Crypto.PassPhrase -> IO (SigningKey ByronKey)
 keygen passphrase =
-  snd <$> Crypto.runSecureRandom (Crypto.safeKeyGen passphrase)
+  ByronSigningKey . snd <$> Crypto.runSecureRandom (Crypto.safeKeyGen passphrase)
```
Comment (dcoutts) [inline cardano-cli/src/Cardano/CLI/Byron/Key.hs:129]: Can't we replace this with `Cardano.Api.generateSigningKey`?

**Review by dcoutts — COMMENTED** [review]
This is looking great. A lot of nice simplifications.

How do we reassure ourselves we've not broken something? We don't have that much automated testing of the Byron parts of the CLI. Perhaps we can check that the `byron-to-shelley/` script still works: it uses the Byron genesis and key generation, and tx submission.

File: cardano-api/src/Cardano/Api/Byron.hs (hunk 3 above)
Comment (Jimbo4350) [reply cardano-api/src/Cardano/Api/Byron.hs:76]: It wasn't but I will move it there.

File: cardano-api/src/Cardano/Api/KeysByron.hs (hunk 5 above)
Comment (Jimbo4350) [reply cardano-api/src/Cardano/Api/KeysByron.hs:78]: @dcoutts How about `PrePBFT` and `PBFT`?

File: cardano-api/src/Cardano/Api/Tx.hs (hunk 6 above)
Comment (Jimbo4350) [reply cardano-api/src/Cardano/Api/Tx.hs:524]: Yep I agree!

File: cardano-cli/src/Cardano/CLI/Byron/Key.hs (hunk 8 above)
Comment (Jimbo4350) [reply cardano-cli/src/Cardano/CLI/Byron/Key.hs:129]: Yep this was my plan, but I got sidetracked.

**Comment by Jimbo4350** [comment]
> How do we reassure ourselves we've not broken something? We don't have that much automated testing of the Byron parts of the CLI. Perhaps we can check that the byron-to-shelley/ script still works: it uses the Byron genesis and key generation, and tx submission.
I've added some deserialisation tests and I will double check all the relevant scripts still work.

File: cardano-api/src/Cardano/Api/KeysByron.hs (hunk 5 above)
Comment (dcoutts) [reply cardano-api/src/Cardano/Api/KeysByron.hs:82]: I think "legacy" vs "non-legacy" or "legacy" vs "modern" is fine. These key formats are not really related to the consensus era (PBFT etc) they just happened to change at the same time, so I don't think PBFT-based naming will be helpful.

So these names are fine.

File: cardano-cli/src/Cardano/CLI/Byron/Run.hs [hunk 9]
```diff
@@ -130,7 +131,10 @@ runMigrateDelegateKeyFrom
         -> ExceptT ByronClientCmdError IO ()
 runMigrateDelegateKeyFrom oldKeyformat oldKey (NewSigningKeyFile newKey) = do
   sk <- firstExceptT ByronCmdKeyFailure $ readByronSigningKey oldKeyformat oldKey
-  firstExceptT ByronCmdHelpersError . ensureNewFileLBS newKey $ serialiseByronWitness sk
+  migratedWitness <- case sk of
+                       LegacyWitness (ByronSigningKeyLegacy sKey) -> return . NonLegacyWitness $ ByronSigningKey sKey
+                       NonLegacyWitness _ -> panic "TODO: Can only migrate from legacy witness"
```
Comment (dcoutts) [inline cardano-cli/src/Cardano/CLI/Byron/Run.hs:136]: We could fail a bit more gracefully here :smiley:

File: cardano-cli/src/Cardano/CLI/Byron/Key.hs [hunk 10]
```diff
@@ -118,38 +103,7 @@ readPaymentVerificationKey (VerificationKeyFile fp) = do
...
-    readOne pr = do
-      hPutStr stdout pr >> hFlush stdout
-      hSetEcho stdout False
-      pp <- SB.hGetLine stdin
-      hSetEcho stdout True
-      hPutStrLn stdout ("" :: String)
-      pure pp
+-- | Generate a cryptographically random signing key.
+keygen :: IO (SigningKey ByronKey)
+keygen = generateSigningKey AsByronKey
```
Comment (dcoutts) [inline cardano-cli/src/Cardano/CLI/Byron/Key.hs:108]: I'd just use this directly now, and remove `keygen`.

File: cardano-api/src/Cardano/Api/KeysByron.hs [hunk 11]
```diff
@@ -54,10 +72,21 @@ import           Cardano.Api.SerialiseTextEnvelope
...
-    proxyToAsType _ = AsByronKey
+class IsByronKey key where
+    byronKeyFormat :: ByronKeyFormat key
+
+data ByronKeyFormat key where
+  ByronLegacyKeyFormat :: ByronKeyFormat ByronKeyLegacy
+  ByronNonLegacyKeyFormat :: ByronKeyFormat ByronKey
+
+data ByronWitness = LegacyWitness (SigningKey ByronKeyLegacy)
+                  | NonLegacyWitness (SigningKey ByronKey)
```
Comment (dcoutts) [inline cardano-api/src/Cardano/Api/KeysByron.hs:85]: I think we should call this `SomeByronSigningKey`, to match the `SomeSigningKey` we have in the CLI.

It's not a witness, it's just a signing key. (Yes signing keys are used to construct witnesses, but they're not a witness)

How about:
```
data SomeByronSigningKey = AByronSigningKey (SigningKey ByronKey)
                         | AByronSigningKeyLegacy (SigningKey ByronKeyLegacy)
```

File: cardano-api/src/Cardano/Api/KeysByron.hs [hunk 12]
```diff
@@ -144,3 +175,119 @@ instance CastVerificationKeyRole ByronKey PaymentKey where
...
+
+instance HasTypeProxy ByronKeyLegacy where
+  data AsType ByronKeyLegacy = AsByronKeyLegacy
+  proxyToAsType _ = AsByronKeyLegacy
+
+instance HasTextEnvelope (VerificationKey ByronKeyLegacy) where
+    textEnvelopeType _ = "PaymentVerificationKeyByronLegacy_ed25519_bip32"
+
+instance HasTextEnvelope (SigningKey ByronKeyLegacy) where
+    textEnvelopeType _ = "PaymentSigningKeyByronLegacy_ed25519_bip32"
```
Comment (dcoutts) [inline cardano-api/src/Cardano/Api/KeysByron.hs:222]: Do we need text envelope instances for them? Don't we just read them in in the old formats and convert to the new? So we'd never be writing out text envelope wrapper with the legacy key format embedded? So if it's not used, perhaps best to remove this instance to help people avoid that mistake. And if it is used, why? :smiley:

File: cardano-api/src/Cardano/Api/SpecialByron.hs [hunk 13]
```diff
@@ -197,4 +199,8 @@ makeByronVote nId (ByronSigningKey sKey) (ByronUpdateProposal proposal) yesOrNo
 toByronLedgertoByronVote :: ByronVote -> Mempool.GenTx ByronBlock
 toByronLedgertoByronVote (ByronVote vote) = Mempool.ByronUpdateVote (recoverVoteId vote) vote
 
-
+byronWitnessToSigningKey :: ByronWitness -> Byron.SigningKey
+byronWitnessToSigningKey bWit =
+  case bWit of
+    LegacyWitness (ByronSigningKeyLegacy sKey) -> sKey
+    NonLegacyWitness (ByronSigningKey sKey) -> sKey
```
Comment (dcoutts) [inline cardano-api/src/Cardano/Api/SpecialByron.hs:206]: When we rename ByronWitness we should rename this too, e.g. simply `toByronSigningKey`, and it'd be fine to put in the ByronKeys module and export via Cardano.Api.Byron.

File: cardano-api/src/Cardano/Api/Tx.hs [hunk 14]
```diff
@@ -521,9 +521,10 @@ makeSignedTransaction witnesses (ShelleyTxBody era txbody txmetadata) =
           (maybeToStrictMaybe txmetadata)
 
 
-makeByronKeyWitness :: NetworkId
+makeByronKeyWitness :: forall key. IsByronKey key
```
Comment (dcoutts) [inline cardano-api/src/Cardano/Api/Tx.hs:524]: Usually we give the forall it's own line, with the class constraints starting on the next line.

**Review by dcoutts — APPROVED** [review]
This is looking great. Good work on the testing too.

Only really comments about naming etc now.

File: cardano-cli/src/Cardano/CLI/Byron/Run.hs (hunk 9 above)
Comment (Jimbo4350) [reply cardano-cli/src/Cardano/CLI/Byron/Run.hs:136]: Yep, this was an oversight.

File: cardano-api/src/Cardano/Api/KeysByron.hs (hunk 12 above)
Comment (Jimbo4350) [reply cardano-api/src/Cardano/Api/KeysByron.hs:222]: Copy pasted without thinking!

Comment (dcoutts) [reply cardano-api/src/Cardano/Api/KeysByron.hs:230]: Ok. Fine.

Maybe we ought to consider relaxing that superclass constraint, but it's fine for now.

**Comment by Jimbo4350** [comment]
bors merge

**Pull request merged**